feat: Redesigned Profile Settings screen for Talawa User Portal

akhilender-bongirwar opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.
-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
This PR includes redesigning of the profile settings of user portal in our application. 
<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**

Fixes #1512 

**Did you add tests for your changes?**
Yes
<!--Yes or No. Note: Add unit tests or automation tests for your code.-->

**Snapshots/Videos:**

https://github.com/PalisadoesFoundation/talawa-admin/assets/112749383/f6ff3f1d-8ab4-4495-b196-f36987e633db


<!--Add snapshots or videos wherever possible.-->

<!--Add link to Talawa-Docs.-->

**Summary**
- Implemented the redesign of user portal settings page as it is given in figma design.
- Ensured there are some components working like the upload picture, d.o.b of the user and the reset changes button.
- Made the page responsive across all the screens.
- All the modified code is covered by tests and ensured no other functionality is affected.
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**
No
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**

<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**
Yes
<!--Yes or No-->

Cioppolo14 commented:
@aashimawadhwa @rishav-jha-mech Can you review this PR?

rishav-jha-mech reviewed (commented):
@akhilender-bongirwar Placeholders texts are missing, also some of those fields will have select option, please fix these things

rishav-jha-mech commented:
Good to go @akhilender-bongirwar, focus on layouts as well, they don't look good on smaller devices. The card with profile pic should be on top once the device width is small enough.

akhilender-bongirwar commented:
@rishav-jha-mech, made necessary modifications. Please check the below video.

https://github.com/PalisadoesFoundation/talawa-admin/assets/112749383/16841342-a1d6-4d1b-b43f-1e85afd67222

rishav-jha-mech commented:
@akhilender-bongirwar 

1. The Copy Profile Link will have some margin on its top
2. Choose file button and its label aren't properly aligned

rishav-jha-mech reviewed (commented):
@akhilender-bongirwar 

1. The Copy Profile Link will have some margin on its top
2. Choose file button and its label aren't properly aligned

noman2002 reviewed (changes requested):
Profile picture should be visible at all the places, not only one place, Instead of "TU" profile image should be visible if profile picture is null then only the name initials will be shown.

pranshugupta54 commented:
@akhilender-bongirwar, could you confirm if the save button is functioning properly? None of your recordings demonstrate whether the data is actually being saved. It seems that the API may require certain fields to be in the correct format, such as Grades.
<img width="180" alt="image" src="https://github.com/PalisadoesFoundation/talawa-admin/assets/76090263/31fcdb10-72a2-4782-bc4e-59de0f762917">

aashimawadhwa commented:
Also @akhilender-bongirwar can you work on the profile settings card? i feel the margins and spacing is off, also the alignment can be improved significantly.
please work on that. 
@rishav-jha-mech this requires your strong assistance.

akhilender-bongirwar commented:
> @akhilender-bongirwar, could you confirm if the save button is functioning properly? None of your recordings demonstrate whether the data is actually being saved. It seems that the API may require certain fields to be in the correct format, such as Grades. <img alt="image" width="180" src="https://private-user-images.githubusercontent.com/76090263/312956523-31fcdb10-72a2-4782-bc4e-59de0f762917.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0ODE3NTEsIm5iZiI6MTcxMDQ4MTQ1MSwicGF0aCI6Ii83NjA5MDI2My8zMTI5NTY1MjMtMzFmY2RiMTAtNzJhMi00NzgyLWJjNGUtNTlkZTBmNzYyOTE3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE1VDA1NDQxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0NWNiZDQ5ZWY5NWE3NzE2ZDk0N2IxY2JhYTdhZWExOGNjOWIzY2YzOWVhY2EwZGQ5YjUxNmU4YWJjMzM0N2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.ys2Dor0gepTDA_7glMZslmAqxoa0pBLrU4o5s5rhU6I">

I apologize that I overlooked this as I was focusing on drafting my proposal at the time. Thank you for the correction. I will modify the code accordingly.

akhilender-bongirwar commented:
https://github.com/PalisadoesFoundation/talawa-admin/assets/112749383/757f30a3-df94-4d7f-9c62-d206c6911a56

- Updated UPDATE_USER_MUTATION to include all the additional fields
- Updated CHECK_AUTH mutation with all the required fields.
- Updated frontend code to utilize these mutations and queries
- Covered the code with tests.

@noman2002 has resolved the profile image error.
 @pranshugupta54, the save button is now functioning correctly.

@aashimawadhwa, the profile card appears satisfactory in desktop/laptop and mobile views, but it seems somewhat dull in tablet view IMO. Could you provide suggestions on how to improve its appearance for tablet users?

pranshugupta54 commented on src/screens/UserPortal/Settings/Settings.tsx:
```diff
@@ -21,148 +24,515 @@ export default function settings(): JSX.Element {
+                        <option value="" disabled>
+                          {t('gradePlaceholder')}
+                        </option>
+                        <option value="GRADE_1">{t('grade1')}</option>
```
Using this approach of embedding all the options directly within the code for the dropdown menu will become increasingly challenging to maintain in the future and is generally considered poor practice.

pranshugupta54 replied:
I've taken a different approach for handling this issue. I created a file in the utilities directory that contains all the enums for each field, along with their corresponding labels and values. These enums are then directly imported into the code, providing a more organized and manageable solution.

Reference: https://github.com/PalisadoesFoundation/talawa-admin/pull/1667/files#diff-2d7b9c41af3deb393284274ccf52ffe702052af5e80d87ed83e9c37606af98cd

akhilender-bongirwar replied:
ok, I will place the values as `gradeOptions` in the `utils/` directory and iterate over them.

akhilender-bongirwar replied:
Cool, you have implemented it in the same way as `countryOptions` were already present.

pranshugupta54 replied:
Hey @palisadoes, I'm considering making a PR to implement the **enums** on Talawa-Admin. Since these enums will be widely used and having separate PRs for each component might lead to conflicts, I'm thinking of creating a single PR for both the enums and the dropdown component. This would streamline the process and reduce potential conflicts. What do you think? Can I proceed with this approach?

akhilender-bongirwar replied:
Hey @pranshugupta54, I have implemented the changes and ensured that translations for the labels are done wherever necessary. I believe it's crucial for us to utilize the multilingual support available in our application.

akhilender-bongirwar commented:
All the checks have been passed  🎉 .

noman2002 reviewed (dismissed):
Changes looks fine to me.

palisadoes commented:
This is an update on the PR merging freeze:

1. In the next few hours we will be merging the `develop-userTypeFix` branch into the `develop` branch. 
1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. This merge will cause some conflicts in your PR.

We decided to do this at the beginning of the weekend to give us all time to adjust PR code and create bug fixes that may arise.

Update your code at or after midnight GMT on the morning of March 23, 2024. (5:30am IST). 

If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

palisadoes commented:
## The PR merging freeze is lifted.

1. We are working on bug fixes that may arise. These should be few and are being tracked here:
    1. https://github.com/orgs/PalisadoesFoundation/projects/24
1. Please update your PRs and local repos. Fix any new merge conflicts that may have occurred.
    
Background:

1. The `develop-userTypeFix` branch was created to fix a long standing design flaw where Admins were Admins of all organizations, not specific ones.
1. The `userType` field has been removed from the User collection and it has been replaced by an `appUserProfileId` field.
    1. This field is `null` if the user isn’t registered to use the apps. This will help people to add users manually during the event checkin process, or if an Admin wants to manually add someone in the Admin dashboard.
    1. When not `null` the `AppUserProfileID` will reference a `AppUserProfile` collection with App related information such as the organizations for which a user may be an Admin.
    1. The updated schema can be found here https://github.com/PalisadoesFoundation/talawa-api/blob/develop-userTypeFix/schema.graphql
    1. This is the parent issue that we have been using to track progress:
        1. https://github.com/PalisadoesFoundation/talawa-api/issues/1965
1. If your PRs have already been approved, request a re-review after fixing the conflicts and refactoring to the new `AppUserProfileID` methodology.

akhilender-bongirwar commented:
https://github.com/PalisadoesFoundation/talawa-admin/assets/112749383/76bbf3d6-a066-4008-9479-804b1fbfd20f

akhilender-bongirwar commented:
@palisadoes, please check

palisadoes commented:
Please fix the failing test

akhilender-bongirwar commented:
Fixed tests. Please check once more : )
